### PR TITLE
IO: Several improvements

### PR DESCRIPTION
**Pull request by eksperimental**
IO Module:
  * Impoves wording in documentation, particularly in the use of :all, :line, count
  * Standardizes arg names to be `line_*`.
  * Adds function guards to check for integers.
  * Unicode is a proper noun, so it should be always capitalized.
  * Adds backticks to args suchs as `device`, `item`, `count`.
  * When functions are "Unicode unsafe", they are indicated in the first paragraph of docs.
  * IO.puts: code simplified removing unnecessary variable.
  * Adds \n to example in IO.gets/2
  * Adds example in IO.chardata_to_string/1 when dealing with strings
  * Renames args in IO.each_stream/2 and IO.each_binstream/2
  * Adds tests for IO.getn/2-3 calling functions with different arities

IO.Steam Module:
  * Adds a note about side effects when dealing with IO streams.

### CODE REVIEW

File: lib/elixir/lib/io.ex [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
   shortcuts to Erlang's `:standard_io` and `:standard_error`.
 
-  The majority of the functions expect char data, i.e. strings or
+  The majority of the functions expect chardata, i.e. strings or
   lists of characters and strings. In case another type is given,
   functions will convert to string via the `String.Chars` protocol
   (as shown in typespecs).
 
-  The functions starting with `bin*` expect iodata as an argument,
+  The functions starting with "bin" expect iodata as an argument,
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: I'm :-1: on this change; I think monospaced font and the `*` are very helpful.

File: lib/elixir/lib/io.ex [hunk 2]
```diff
@@ -47,8 +47,11 @@ defmodule IO do
   end
 
   @doc """
-  Reads `count` characters from the IO device, a whole
-  `:line` or the whole device with `:all`.
+  Reads the IO `device`.
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: I'd shove a "from" in there:

> Reads from the IO `device`.

or something like this.

File: lib/elixir/lib/io.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: backticks are fine, but I feel it is redundant to use "starting with" and "*"  together

Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Good point. I'd leave the backticks then, and let someone else decide what to do with the `*` 😄

File: lib/elixir/lib/io.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: I have replaced it everywhere it omited the from

File: lib/elixir/lib/io.ex [hunk 3]
```diff
@@ -171,14 +178,13 @@ defmodule IO do
   end
 
   @doc """
-  Writes the argument to the device, similar to `write/2`,
-  but adds a newline at the end. The argument is expected
-  to be a chardata.
+  Writes the given argument to the `device`, similar to `write/2`,
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Maybe we can say

> Writes the given `item` to the [...]

Wdyt?

File: lib/elixir/lib/io.ex [hunk 4]
```diff
@@ -171,14 +178,13 @@ defmodule IO do
...
+  but adds a newline at the end.
+
   """
   @spec puts(device, chardata | String.Chars.t) :: :ok
   def puts(device \\ group_leader(), item) do
-    erl_dev = map_dev(device)
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: This variable serves as some kind of documentation IMO; I'm not sure if we should refactor it out (or if that change belongs to a style/docs PR).

File: lib/elixir/lib/io.ex [hunk 5]
```diff
@@ -201,7 +207,7 @@ defmodule IO do
   end
 
   @doc """
-  Inspects the item with options using the given device.
+  Inspects the `item` with options using the given `device`.
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Maybe

> Inspects the `item` **according to the given options** using the given `device`.

?

File: lib/elixir/lib/io.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: Everywhere else in the code map_dev/1 is used directly. you can see in the function above.

File: lib/elixir/lib/io.ex [hunk 6]
```diff
@@ -230,9 +238,9 @@ defmodule IO do
   """
   @spec getn(chardata | String.Chars.t, pos_integer) :: chardata | nodata
   @spec getn(device, chardata | String.Chars.t) :: chardata | nodata
-  def getn(prompt, count \\ 1)
+  def getn(device, count \\ 1)
 
-  def getn(prompt, count) when is_integer(count) do
+  def getn(prompt, count) when is_integer(count) and count > 0 do
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Maybe we should change the first argument here to `device` as well? Just for consistency, and also we're changing this line anyways so the blame won't be affected.

File: lib/elixir/lib/io.ex [hunk 7]
```diff
@@ -301,37 +311,44 @@ defmodule IO do
 
   """
   @spec stream(device, :line | pos_integer) :: Enumerable.t
-  def stream(device, line_or_codepoints) do
+  def stream(device, line_or_codepoints)
+  when line_or_codepoints == :line
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: In Elixir, `when` is usually indented two levels (4 spaces) more than the thing it belongs to (`def` here).

File: lib/elixir/lib/io.ex [hunk 8]
```diff
@@ -301,37 +311,44 @@ defmodule IO do
...
+  or is_integer(line_or_codepoints) and line_or_codepoints > 0 do
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Maybe this would look cleaner if we put parens around the `and` expression? Just to make precedence very clear.

File: lib/elixir/lib/io.ex [hunk 9]
```diff
@@ -301,37 +311,44 @@ defmodule IO do
...
-  Finally, do not use this function on IO devices in unicode
+  Finally, do not use this function on IO devices in Unicode
   mode as it will return the wrong result.
+
   """
   @spec binstream(device, :line | pos_integer) :: Enumerable.t
-  def binstream(device, line_or_bytes) do
+  def binstream(device, line_or_bytes)
+  when line_or_bytes == :line
+  or is_integer(line_or_bytes) and line_or_bytes > 0 do
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Same as above.

File: lib/elixir/lib/io.ex [hunk 10]
```diff
@@ -301,37 +311,44 @@ defmodule IO do
...
     IO.Stream.__build__(map_dev(device), true, line_or_bytes)
   end
 
   @doc """
   Converts chardata (a list of integers representing codepoints,
   lists and strings) into a string.
 
-  In case the conversion fails, it raises a `UnicodeConversionError`.
-  If a string is given, returns the string itself.
+
```
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: ✂️

File: lib/elixir/lib/io.ex (hunk 4 above)
Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: Ah, good point again :) Sorry for the false alarm then!

File: lib/elixir/lib/io.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: I think it's clearer, but .. too many "given"s

File: lib/elixir/lib/io.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: sure... but i realized the origina version had too many "given"s
how about this?
> Writes `item` as a binary to the given `device`.

File: lib/elixir/lib/io.ex (hunk 8 above)
Comment (ericmj) [inline lib/elixir/lib/io.ex]: It's also convention to leave operators at the end of the line instead of starting new lines with them.

File: lib/elixir/lib/io.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: I thought of that too while reviewing the code, and by doing that we can get rid of the body less function as well..
but I think whoever wrote that line is trying to make it clear that prompt is reading from stdin,
or that's what I understand from it.

Comment (whatyouhide) [inline lib/elixir/lib/io.ex]: I think in this case, a comment may be a better tool to explain that.

Comment (eksperimental) [inline lib/elixir/lib/io.ex]: I reverted all the changes in the IO.getn/2-3 functions.
Since I have refactored the code, and I think it deserves a separate PR.

File: lib/elixir/lib/io.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: finally changed to:
> Inspects `item` according to the given options using the IO `device`.

please let me know what you think @whatyouhide

File: lib/elixir/lib/io.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/io.ex]: ok. I have analyzed the situation and refactoring didn't work.
What we have is the best we can do. I will leave a note in the PR comments

**Comment by eksperimental** [comment]
this is ready for review.
I have added a few final changes, including some tests.

thank you @whatyouhide for all your suggestions.
regarding changing the arg name for IO.getn/2,
It's ok as it is. As this function is trying work as :erlang.get_chars/2
See the arguments there
http://erldocs.com/18.3/stdlib/io.html?i=0&search=get_chars#get_chars/2
* `get_chars(Prompt, Count) -> Data | server_no_data()`
* `get_chars(IoDevice, Prompt, Count) -> Data | server_no_data()` 

I tried to refactor it and it didn't work because setting default values in the function definition of IO.getn/3 were overlapping IO.getn/2

**Comment by eksperimental** [comment]
Another comment is that we are defining two typespecs for IO.getn/2
```elixir
  @spec getn(chardata | String.Chars.t, pos_integer) :: chardata | nodata
  @spec getn(device, chardata | String.Chars.t) :: chardata | nodata
```
And that can give a Dyalizer warning. I'm  not sure if we should removing the second one or leave it as it is.

**Comment by ericmj** [comment]
> And that can give a Dyalizer warning. I'm not sure if we should removing the second one or leave it as it is.

That shouldn't produce warnings, removing it probably will though.

File: lib/elixir/lib/io.ex [hunk 11]
```diff
@@ -214,45 +220,53 @@ defmodule IO do
...
   @spec getn(device, chardata | String.Chars.t) :: chardata | nodata
   def getn(prompt, count \\ 1)
 
-  def getn(prompt, count) when is_integer(count) do
+  def getn(prompt, count) when is_integer(count) and count > 0 do
     getn(group_leader, prompt, count)
   end
 
+  # This a catch-all function to call IO.getn/3
   def getn(device, prompt) do
```
Comment (ericmj) [inline lib/elixir/lib/io.ex]: Is this clause valid for integers less than 0? If not the added guard above won't really help unless you also add `when not is_integer(prompt)` here.

**Comment by eksperimental** [comment]
@ericmj could you please explain what is the difference with the one that was removed from the Range module. What I understood was that having two overlapping specs would give a warning in Dyalizer. thanks

**Comment by ericmj** [comment]
If they are overlapping Dialyzer will warn, but where is it overlapping?

**Comment by josevalim** [comment]
This looks great, thank you. Let's wait for the last few notes and merge it. :)

**Comment by eksperimental** [comment]
@ericmj I have addressed the issue regarding the guard. good catch!

this is ready for merge

**Comment by eksperimental** [comment]
@ericmj I keep thinking of specs as functions, not as specs.. what's why I saw the overlapping

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
